### PR TITLE
Add `best_k_metrics` parameter to the `ModelCheckpoint`

### LINKED ISSUE #20321: best-k-metrics in ModelCheckpoint
### Description & Motivation

As part of the ModelCheckpoint, it would be nice to have access to all the best metrics, not just the monitor one. You can check an example in #feature/best-k-metrics. Implementation here: [https://github.com/gonzachiar/pytorch-lightning/commit/485c367fa37093749e5f9f59305c56d2ef284198](https://github.com/gonzachiar/pytorch-lightning/commit/485c367fa37093749e5f9f59305c56d2ef284198)

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gonzachiar**
## What does this PR do?

Adds a parameter to save all the metrics from the best model.

Fixes #20321

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [X] Is this pull request ready for review? (if not, please submit in draft mode)
- [X] Check that all items from **Before submitting** are resolved
- [X] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20457.org.readthedocs.build/en/20457/

### CODE REVIEW

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -241,9 +249,10 @@ def __init__(
         self._last_global_step_saved = 0  # no need to save when no steps were taken
         self._last_time_checked: Optional[float] = None
         self.current_score: Optional[Tensor] = None
-        self.best_k_models: Dict[str, Tensor] = {}
+        self.best_k_models: Dict[str, Dict[str, Tensor | Dict[str, Tensor]]] = {}
```
Comment (lantiga) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:252]: this may be easier to read
`Dict[str, Dict[str, Tensor]] | Dict[str, Dict[str, Dict[str, Tensor]]]`
 but ultimately we'd be better off defining a type alias

more importantly, we need to avoid breaking backward compatibility here
so whatever code relies on `best_k_models` being `Dict[str, Tensor]` today needs to keep working

I suggest we just limit ourselves to track `best_model_metrics` and not mess with `best_k_models`, or use a separate private attribute

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -523,7 +534,9 @@ def check_monitor_top_k(self, trainer: "pl.Trainer", current: Optional[Tensor] =
             return True
 
         monitor_op = {"min": torch.lt, "max": torch.gt}[self.mode]
-        should_update_best_and_save = monitor_op(current, self.best_k_models[self.kth_best_model_path])
+        should_update_best_and_save = monitor_op(
+            current, cast(Tensor, self.best_k_models[self.kth_best_model_path]["score"])
```
Comment (lantiga) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:538]: this will stay as in the original if we avoid changing `best_k_models`

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 3]
```diff
@@ -706,6 +706,7 @@ def test_model_checkpoint_save_last_none_monitor(tmp_path, caplog):
     assert checkpoint_callback.best_model_path == str(tmp_path / "epoch=1-step=20.ckpt")
     assert checkpoint_callback.last_model_path == str(tmp_path / "last.ckpt")
     assert checkpoint_callback.best_model_score is None
+    assert checkpoint_callback.best_model_metrics is None
```
Comment (lantiga) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:709]: we need to add tests that exercise the new code

**Review by lantiga — COMMENTED** [review]
Thanks for the contribution! Happy to merge after we ensure full backward compatibility and test coverage

**Comment by lantiga** [comment]
@gonzachiar I'm wrapping up the last few PRs for the release, do you have time to push this through in the next couple of days?

**Comment by gonzachiar** [comment]
> @gonzachiar I'm wrapping up the last few PRs for the release, do you have time to push this through in the next couple of days?

Hey, I will work on this during the weekend. Hopes that helps!

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py (hunk 3 above)
Comment (Borda) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py:712]: @gonzachiar, you have only test for the actual scenarion when the newly added attribute is empty, but we need to add a test to validate that your addition works as expected

**Comment by deependujha** [comment]
hi @gonzachiar, not sure what exactly happened over here in last few minutes. Why so many commits and many irrelevant changes.

Can you please narrow down changes to be issue specific? Thanks.

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 4]
```diff
@@ -32,7 +32,6 @@
 from torch import optim
 from torch.utils.data.dataloader import DataLoader
 
-import lightning.pytorch as pl
```
Comment (deependujha) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:35]: Why was it removed?

Comment (deependujha) [reply tests/tests_pytorch/checkpointing/test_model_checkpoint.py:35]: It causes `pre-commit failure`.

**Comment by gonzachiar** [comment]
> hi @gonzachiar, not sure what exactly happened over here in last few minutes. Why so many commits and many irrelevant changes.
> 
> Can you please narrow down changes to be issue specific? Thanks.

hi, sorry this PR was really outdated and I'm trying to solve all the conflicts. Please let me a couple of minutes and I will fix it.

**Comment by gonzachiar** [comment]
@deependujha that should be good, solicited a review again :)
